# Prediction of H1N1 and Seasonal Flu Vaccinations
## 1. Introduction


### 1.1 Problem Statement
### 1.2 Main Objective

## 1.3 Metric for Success
## 1.4 Experimental Design

## 1.5 Data Description
The datasets used for this project were downloaded from [Driven Data]("https://www.drivendata.org/competitions/66/flu-shot-learning/page/210/). The original data source is the [National 2009 H1N1 Flu Survey (NHFS)](https://webarchive.loc.gov/all/20140511165905/http://www.cdc.gov/nchs/nis/data_files_h1n1.htm) and it contains information on the social,economic and demographic backgrounds of the respondents as well as their opinions on the H1N1 and seasonal flu vaccines. The datasets have been divided into the training set features, the training set labels and the test set features. The training data has 26707 rows and 36 columns. The information contained with the columns is as follows as described by the data [dictionary](https://www.drivendata.org/competitions/66/flu-shot-learning/page/211/):

|No.| Column    | Description|
|---| ---       | ---         |
|1| `respondent_id` | Unique and random identifier for the respondents|
|2|`h1n1_concern`| Level of concern about H1N1 flu with 0 being not concerned at all and 3 being very concerned|
|3|`h1n1_knowledge`| Level of knowledge about H1N1 with 0 being no knowledge and 2 being a lot of knowledge|
|4|`behavioral_antiviral_meds`| Has taken any antiviral medication (0-no,1-yes)|
|5|`behavioral_avoidance`| Has avoided close contact with anyone with flu-like symptoms (0-no,1-yes)|
|6|`behavioral_face_mask`| Has bought a face mask (0-no,1-yes)|
|7|`behavioral_wash_hands`| Has frequently washed hands or used hand sanitizer (0-no,1-yes)|
|8|`behavioral_large_gatherings`| Has reduced time at large gatherings (0-no,1-yes)|
|9|`behavioral_outside_home`| Has reduced contact with people outside of own household (0-no,1-yes)|
|10|`behavioral_touch_face`| Has avoided touching eyes, nose or mouth (0-no,1-yes)|
|11|`doctor_recc_h1n1`| H1N1 flu vaccine was recommended by doctor (0-no,1-yes)|
|12|`doctor_recc_seasonal`| H1N1 flu vaccine was recommended by doctor (0-no,1-yes)|
|13|`chronic_med_condition`| Has any of the following chronic conditions: asthma or any lung condition, a heart condition, a kidney condition, sickle cell anaemia or any other anaemia, a neurological or neouromuscular condition, a liver condition, or a weakened immune system as a result of a chronic illness or medicines taken for a chronic illness (0-no,1-yes)|
|14|`child_under_6_months`| Has regular close contact with a child under the age of six months (0-no,1-yes)|
|15|`health_worker`| Is a healthcare worker (0-no,1-yes)|
|16|`health_insurance`| Has health insurance (0-no,1-yes)|
|17|`opinion_h1n1_vacc_effective`| Respondent's opinion on the efficacy of the vaccine with 1 being not at all effective and 5 being very effective|
|18|`opinion_h1n1_risk`| Respondent's opinion about risk of getting sick with H1N1 flu without vaccine with 1 being very low and 5 being very high|
|19|`opinion_h1n1_sick_from_vacc`| Respondent's worry of getting sick from H1N1 vaccine with 1 being not worried at all and 5 being very worried|
|20|`opinion_seas_vacc_effective`| Respondent's opinion about seasonal flu vaccine effectiveness with 1 being not effective at all and 5 being very effective|
|21|`opinion_seas_risk`| Respondent's opinion about risk of getting sick with seasonal flu without vaccine with 1 being very low and 5 being very high|
|22|`opinion_seas_sick_from_vacc`| Respondent's worry of getting sick from taking seasonal flu vaccine with 1 being not worried at all and 5 being very worried|
|23|`age_group`| Age group of respondents|
|24|`education`| Self-reported educational level|
|25|`race`| Race of respondent|
|26|`sex`| Sex of respondent|
|27|`income_poverty`| Household annual income of respondent with respect to 2008 Census poverty thresholds|
|28|`marital status`| Marital status of respondent|
|29|`rent_or_own`| Housing situation of respondent|
|30|`employment_status`| Employment status of respondent|
|31|`hhs_geo_region`| Respondent's residence using a 10-region geographic classification defined by the U.S. Dept. of Health and Human Services. Values are represented as short random character strings|
|32|`census_msa`| Respondent's residence within metropolitan statistical areas (MSA) as defined by the U.S. Census|
|33|`household_adults`| Number of *other* adults in the household, top-coded to 3|
|34|`household_children`| Number of children in the household, top-coded to 3|
|35|`employment_industry`| Type of industry respondent is employed in. Values are represented as short random character strings|
|36|`employment_occupation`| Type of occupation of respondent.Values are represented as short random character strings| 

## 2. Reading the Data

In [2]:
# importing relevant libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

from sklearn.model_selection import train_test_split()

In [3]:
# setting maximum number of columns to be displayed
pd.set_option('display.max_columns',100)

In [4]:
# loading training features into a dataframe and previewing first five rows
train_features_df = pd.read_csv("Dataset/training_set_features.csv",index_col = "respondent_id")
train_features_df.head()

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
0                       1.0             0.0                        0.0   
1                       3.0             2.0                        0.0   
2                       1.0             1.0                        0.0   
3                       1.0             1.0                        0.0   
4                       2.0             1.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
0                               0.0                   0.0   
1                               1.0                   0.0   
2                               1.0                   0.0   
3                               1.0                   0.0   
4                               1.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
0                                0.0                          0.0   
1                                1.0                          0.0   
2                                0.0                          0.0   
3                                1.0                          1.0   
4                                1.0                          1.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
0                                  1.0                    1.0   
1                                  1.0                    1.0   
2                                  0.0                    0.0   
3                                  0.0                    0.0   
4                                  0.0                    1.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
0                           0.0                   0.0                    0.0   
1                           0.0                   0.0                    0.0   
2                           NaN                   NaN                    1.0   
3                           0.0                   1.0                    1.0   
4                           0.0                   0.0                    0.0   

               child_under_6_months  health_worker  health_insurance  \
respondent_id                                                          
0                               0.0            0.0               1.0   
1                               0.0            0.0               1.0   
2                               0.0            0.0               NaN   
3                               0.0            0.0               NaN   
4                               0.0            0.0               NaN   

               opinion_h1n1_vacc_effective  opinion_h1n1_risk  \
respondent_id                                                   
0                                      3.0                1.0   
1                                      5.0                4.0   
2                                      3.0                1.0   
3                                      3.0                3.0   
4                                      3.0                3.0   

               opinion_h1n1_sick_from_vacc  opinion_seas_vacc_effective  \
respondent_id                                                             
0                                      2.0                          2.0   
1                                      4.0                          4.0   
2                                      1.0                          4.0   
3                                      5.0                          5.0   
4                                      2.0                          3.0   

               opinion_seas_risk  opinion_seas_sick_from_vacc      age_group  \


In [5]:
# loading training labels into a dataframe and previewing first five rows
train_label_df = pd.read_csv("Dataset/training_set_labels.csv",index_col = "respondent_id")
train_label_df.head()

h1n1_vaccine  seasonal_vaccine
respondent_id                                
0                         0                 0
1                         0                 1
2                         0                 0
3                         0                 1
4                         0                 0

In [6]:
# assert that the respondent_id in features and labels column are the same in order to merge
assert train_features_df.index.equals(train_label_df.index)

In [7]:
# joining the two dataframes
data = train_features_df.join(train_label_df)
data.shape

(26707, 37)

## 3. Assessing the Data

In [14]:
# previewing the last five rows
data.tail()

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
26702                   2.0             0.0                        0.0   
26703                   1.0             2.0                        0.0   
26704                   2.0             2.0                        0.0   
26705                   1.0             1.0                        0.0   
26706                   0.0             0.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
26702                           1.0                   0.0   
26703                           1.0                   0.0   
26704                           1.0                   1.0   
26705                           0.0                   0.0   
26706                           1.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
26702                            0.0                          0.0   
26703                            1.0                          0.0   
26704                            1.0                          1.0   
26705                            0.0                          0.0   
26706                            0.0                          0.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
26702                              1.0                    0.0   
26703                              0.0                    0.0   
26704                              0.0                    1.0   
26705                              0.0                    NaN   
26706                              0.0                    0.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
26702                       0.0                   0.0                    0.0   
26703                       1.0                   1.0                    0.0   
26704                       0.0                   0.0                    0.0   
26705                       0.0                   0.0                    0.0   
26706                       0.0                   0.0                    0.0   

               child_under_6_months  health_worker  health_insurance  \
respondent_id                                                          
26702                           0.0            0.0               NaN   
26703                           0.0            1.0               1.0   
26704                           0.0            0.0               NaN   
26705                           0.0            0.0               0.0   
26706                           0.0            0.0               1.0   

               opinion_h1n1_vacc_effective  opinion_h1n1_risk  \
respondent_id                                                   
26702                                  3.0                1.0   
26703                                  4.0                2.0   
26704                                  4.0                4.0   
26705                                  3.0                1.0   
26706                                  5.0                1.0   

               opinion_h1n1_sick_from_vacc  opinion_seas_vacc_effective  \
respondent_id                                                             
26702                                  1.0                          5.0   
26703                                  2.0                          5.0   
26704                                  2.0                          5.0   
26705                                  2.0                          2.0   
26706                                  1.0                          5.0   

               opinion_seas_risk  opinion_seas_sick_from_vacc      age_group  \


In [15]:
# previewing number of rows and columns
data.shape

(26707, 37)

In [16]:
# previewing the datatypes
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26707 entries, 0 to 26706
Data columns (total 37 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   h1n1_concern                 26615 non-null  float64
 1   h1n1_knowledge               26591 non-null  float64
 2   behavioral_antiviral_meds    26636 non-null  float64
 3   behavioral_avoidance         26499 non-null  float64
 4   behavioral_face_mask         26688 non-null  float64
 5   behavioral_wash_hands        26665 non-null  float64
 6   behavioral_large_gatherings  26620 non-null  float64
 7   behavioral_outside_home      26625 non-null  float64
 8   behavioral_touch_face        26579 non-null  float64
 9   doctor_recc_h1n1             24547 non-null  float64
 10  doctor_recc_seasonal         24547 non-null  float64
 11  chronic_med_condition        25736 non-null  float64
 12  child_under_6_months         25887 non-null  float64
 13  health_worker   

In [13]:
# previewing summary statistics of columns
data.describe()

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
count  26615.000000    26591.000000               26636.000000   
mean       1.618486        1.262532                   0.048844   
std        0.910311        0.618149                   0.215545   
min        0.000000        0.000000                   0.000000   
25%        1.000000        1.000000                   0.000000   
50%        2.000000        1.000000                   0.000000   
75%        2.000000        2.000000                   0.000000   
max        3.000000        2.000000                   1.000000   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
count          26499.000000          26688.000000           26665.000000   
mean               0.725612              0.068982               0.825614   
std                0.446214              0.253429               0.379448   
min                0.000000              0.000000               0.000000   
25%                0.000000              0.000000               1.000000   
50%                1.000000              0.000000               1.000000   
75%                1.000000              0.000000               1.000000   
max                1.000000              1.000000               1.000000   

       behavioral_large_gatherings  behavioral_outside_home  \
count                  26620.00000             26625.000000   
mean                       0.35864                 0.337315   
std                        0.47961                 0.472802   
min                        0.00000                 0.000000   
25%                        0.00000                 0.000000   
50%                        0.00000                 0.000000   
75%                        1.00000                 1.000000   
max                        1.00000                 1.000000   

       behavioral_touch_face  doctor_recc_h1n1  doctor_recc_seasonal  \
count           26579.000000      24547.000000          24547.000000   
mean                0.677264          0.220312              0.329735   
std                 0.467531          0.414466              0.470126   
min                 0.000000          0.000000              0.000000   
25%                 0.000000          0.000000              0.000000   
50%                 1.000000          0.000000              0.000000   
75%                 1.000000          0.000000              1.000000   
max                 1.000000          1.000000              1.000000   

       chronic_med_condition  child_under_6_months  health_worker  \
count           25736.000000          25887.000000   25903.000000   
mean                0.283261              0.082590       0.111918   
std                 0.450591              0.275266       0.315271   
min                 0.000000              0.000000       0.000000   
25%                 0.000000              0.000000       0.000000   
50%                 0.000000              0.000000       0.000000   
75%                 1.000000              0.000000       0.000000   
max                 1.000000              1.000000       1.000000   

       health_insurance  opinion_h1n1_vacc_effective  opinion_h1n1_risk  \
count       14433.00000                 26316.000000       26319.000000   
mean            0.87972                     3.850623           2.342566   
std             0.32530                     1.007436           1.285539   
min             0.00000                     1.000000           1.000000   
25%             1.00000                     3.000000           1.000000   
50%             1.00000                     4.000000           2.000000   
75%             1.00000                     5.000000           4.000000   
max             1.00000                     5.000000           5.000000   

       opinion_h1n1_sick_from_vacc  opinion_seas_vacc_effective  \
count                 26312.000000                 26245.000000   
mean                      2.357670                     4.025986   
std        

In [18]:
# checking for duplicates
data.duplicated().sum()

0

In [21]:
# checking for missing values
missing = data.isna().sum().to_frame().sort_values(0,ascending = False)
missing

0
employment_occupation        13470
employment_industry          13330
health_insurance             12274
income_poverty                4423
doctor_recc_h1n1              2160
doctor_recc_seasonal          2160
rent_or_own                   2042
employment_status             1463
marital_status                1408
education                     1407
chronic_med_condition          971
child_under_6_months           820
health_worker                  804
opinion_seas_sick_from_vacc    537
opinion_seas_risk              514
opinion_seas_vacc_effective    462
opinion_h1n1_sick_from_vacc    395
opinion_h1n1_vacc_effective    391
opinion_h1n1_risk              388
household_adults               249
household_children             249
behavioral_avoidance           208
behavioral_touch_face          128
h1n1_knowledge                 116
h1n1_concern                    92
behavioral_large_gatherings     87
behavioral_outside_home         82
behavioral_antiviral_meds       71
behavioral_wash_hands           42
behavioral_face_mask            19
sex                              0
race                             0
age_group                        0
hhs_geo_region                   0
census_msa                       0
h1n1_vaccine                     0
seasonal_vaccine                 0

In [36]:
# checking for distribution of target ie h1n1_vaccine and seasonal vaccine
print(f"Count of people vaccinated against H1N1: \n{data.h1n1_vaccine.value_counts()}\n")
print(f"Percentage of people vaccinated against H1N1: \n{data.h1n1_vaccine.value_counts(normalize = True)}\n")
print(f"Count of people vaccinated against seasonal flu: \n{data.seasonal_vaccine.value_counts()}\n")
print(f"Percentage of people vaccinated against seasonal flu: \n{data.seasonal_vaccine.value_counts(normalize = True)}\n")

Count of people vaccinated against H1N1: 
0    21033
1     5674
Name: h1n1_vaccine, dtype: int64

Percentage of people vaccinated against H1N1: 
0    0.787546
1    0.212454
Name: h1n1_vaccine, dtype: float64

Count of people vaccinated against seasonal flu: 
0    14272
1    12435
Name: seasonal_vaccine, dtype: int64

Percentage of people vaccinated against seasonal flu: 
0    0.534392
1    0.465608
Name: seasonal_vaccine, dtype: float64



### Key Observations
* Most columns with float datatype are categorical in nature with values ranging from 0-5.
* There are no duplicate rows
* Almost all columns have missing values
* The columns, hhs_geo_region, employment_industry and employment_occupation are encoded with random strings, possibly to anonymise the data, they will need to be replaced with numbers for readability
* There is a class imbalance issue in the h1n1_vaccine column, a model predicting only 0s (person is not vaccinated) would be correct 78% of the time

## 4. Data Cleaning
This section aims to deal apply data cleaning techniques in order to prepare the data for exploratory data analysis. It addresses missing data, wrong data types, invalid datatypes and any feature engineering needed.

### 4.1 Missing values
##### `employment_occupation` and `employment_industry` column
These columns hava 13470 and 13330 missing values which account for 50% of the dataset. The missing values can be replaced with word "missing"

In [39]:
# replacing missing values in the mentioned columns
data.fillna({"employment_occupation":"missing","employment_industry":"missing"},inplace = True)

##### `health_insurance` column
This column has 12274 missing values. It can be assumed that the missing values are for people without health insurance that replacing missing values with 0

In [42]:
# replacing missing values in this column
data.fillna({"health_insurance": 0},inplace = True)

##### `employment_status` column
This column has 1463

In [51]:
data.income_poverty.value_counts()

<= $75,000, Above Poverty    12777
> $75,000                     6810
Below Poverty                 2697
Name: income_poverty, dtype: int64

In [45]:
4423/26707

0.16561201183210394

In [49]:
data[data.doctor_recc_h1n1.isnull()].index.equals(data[data.doctor_recc_seasonal.isnull()].index)

True

In [50]:
data[data.doctor_recc_h1n1.isnull()]

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
2                       1.0             1.0                        0.0   
76                      2.0             1.0                        0.0   
90                      1.0             1.0                        0.0   
94                      2.0             0.0                        1.0   
97                      2.0             2.0                        0.0   
...                     ...             ...                        ...   
26675                   2.0             2.0                        0.0   
26679                   3.0             1.0                        0.0   
26682                   2.0             2.0                        0.0   
26692                   1.0             1.0                        0.0   
26695                   0.0             0.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
2                               1.0                   0.0   
76                              NaN                   0.0   
90                              1.0                   0.0   
94                              0.0                   0.0   
97                              1.0                   0.0   
...                             ...                   ...   
26675                           1.0                   0.0   
26679                           1.0                   0.0   
26682                           1.0                   0.0   
26692                           0.0                   0.0   
26695                           0.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
2                                0.0                          0.0   
76                               0.0                          0.0   
90                               1.0                          0.0   
94                               1.0                          0.0   
97                               1.0                          0.0   
...                              ...                          ...   
26675                            1.0                          0.0   
26679                            0.0                          1.0   
26682                            1.0                          0.0   
26692                            1.0                          0.0   
26695                            0.0                          0.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
2                                  0.0                    0.0   
76                                 0.0                    0.0   
90                                 0.0                    1.0   
94                                 1.0                    1.0   
97                                 0.0                    1.0   
...                                ...                    ...   
26675                              0.0                    1.0   
26679                              1.0                    1.0   
26682                              0.0                    1.0   
26692                              0.0                    1.0   
26695                              0.0                    0.0   

               doctor_recc_h1n1  doctor_recc_seasonal  chronic_med_condition  \
respondent_id                                                                  
2                           NaN                   NaN                    1.0   
76                          NaN                   NaN                    1.0   
90                          NaN                   NaN                    0.0   
94                          NaN                   NaN                    0.0   
97                          NaN   

In [43]:
data.isna().sum().to_frame().sort_values(0,ascending = False)

0
income_poverty               4423
doctor_recc_seasonal         2160
doctor_recc_h1n1             2160
rent_or_own                  2042
employment_status            1463
marital_status               1408
education                    1407
chronic_med_condition         971
child_under_6_months          820
health_worker                 804
opinion_seas_sick_from_vacc   537
opinion_seas_risk             514
opinion_seas_vacc_effective   462
opinion_h1n1_sick_from_vacc   395
opinion_h1n1_vacc_effective   391
opinion_h1n1_risk             388
household_adults              249
household_children            249
behavioral_avoidance          208
behavioral_touch_face         128
h1n1_knowledge                116
h1n1_concern                   92
behavioral_large_gatherings    87
behavioral_outside_home        82
behavioral_antiviral_meds      71
behavioral_wash_hands          42
behavioral_face_mask           19
age_group                       0
race                            0
sex                             0
health_insurance                0
hhs_geo_region                  0
census_msa                      0
employment_industry             0
employment_occupation           0
h1n1_vaccine                    0
seasonal_vaccine                0

In [22]:
data.corr()

h1n1_concern  h1n1_knowledge  \
h1n1_concern                     1.000000        0.062522   
h1n1_knowledge                   0.062522        1.000000   
behavioral_antiviral_meds        0.090301       -0.010836   
behavioral_avoidance             0.233713        0.088839   
behavioral_face_mask             0.156032        0.030050   
behavioral_wash_hands            0.293565        0.089785   
behavioral_large_gatherings      0.255031       -0.048898   
behavioral_outside_home          0.246424       -0.068219   
behavioral_touch_face            0.248016        0.086162   
doctor_recc_h1n1                 0.150206        0.093829   
doctor_recc_seasonal             0.135567        0.071920   
chronic_med_condition            0.095162       -0.022602   
child_under_6_months             0.049574        0.021598   
health_worker                    0.033562        0.170252   
health_insurance                -0.003655        0.119408   
opinion_h1n1_vacc_effective      0.239604        0.121009   
opinion_h1n1_risk                0.377033        0.073086   
opinion_h1n1_sick_from_vacc      0.360070       -0.019501   
opinion_seas_vacc_effective      0.235136        0.086103   
opinion_seas_risk                0.333990        0.076825   
opinion_seas_sick_from_vacc      0.225731       -0.062483   
household_adults                -0.015658        0.024874   
household_children               0.050525        0.050583   
h1n1_vaccine                     0.121929        0.117951   
seasonal_vaccine                 0.154828        0.120152   

                             behavioral_antiviral_meds  behavioral_avoidance  \
h1n1_concern                                  0.090301              0.233713   
h1n1_knowledge                               -0.010836              0.088839   
behavioral_antiviral_meds                     1.000000              0.049247   
behavioral_avoidance                          0.049247              1.000000   
behavioral_face_mask                          0.146261              0.064946   
behavioral_wash_hands                         0.064119              0.338130   
behavioral_large_gatherings                   0.106287              0.227675   
behavioral_outside_home                       0.127679              0.220348   
behavioral_touch_face                         0.070868              0.335335   
doctor_recc_h1n1                              0.051235              0.068145   
doctor_recc_seasonal                          0.030909              0.074088   
chronic_med_condition                         0.008465              0.039435   
child_under_6_months                          0.028788             -0.000414   
health_worker                                 0.009465              0.001180   
health_insurance                             -0.063988              0.032662   
opinion_h1n1_vacc_effective                   0.030211              0.112300   
opinion_h1n1_risk                             0.105163              0.117851   
opinion_h1n1_sick_from_vacc                   0.079091              0.131405   
opinion_seas_vacc_effective                   0.015003              0.119554   
opinion_seas_risk                             0.085315              0.129504   
opinion_seas_sick_from_vacc                   0.084305              0.082942   
household_adults                              0.044900              0.019122   
household_children                            0.084822              0.040328   
h1n1_vaccine                                  0.040608              0.047690   
seasonal_vaccine                              0.006277              0.076395   

                             behavioral_face_mask  behavioral_wash_hands  \
h1n1_concern                             0.156032               0.293565   
h1n1_knowledge                           0.030050               0.089785   
behavioral_antiviral_meds                0.146261               0.064119   
behavioral_avoidance                     0.064946        